src/alloc.rs

```diff
@@ -65,6 +67,119 @@ impl AllocLock {
+/// removal.
+///
+/// Unlike the `Vec<T>` type available in the standard library, `MaybeVec<T>`
+/// does not shift remaining elements to the left after removing an item.
+/// Instead, a `None` value is swapped in its place, effectively leaving a hole
+/// inside the vector.
+///
+/// Accessing elements in a `MaybeVec<T>`, however, can be slower than a
+/// `Vec<T>`, as elements are no longer strictly contiguous; fragmentation will
+/// occur over time as removed elements increase the distance between those
+/// which remain.
+///
+/// The API to `MaybeVec<T>` can be thought of as a heavily stripped down
+/// version of `Vec<T>`. Its use-case is highly specific, and intended only to
+/// be used internally as part of the Collector implementation. For this reason,
+/// its allocation is not tracked by the collector. Do not store any values
+/// inside which you intend the GC to keep alive.
+pub(crate) struct MaybeVec<T> {
+    buf: RawVec<Option<T>, System>,
+    len: usize,
+}
+
+impl<T> MaybeVec<T> {
+    #[inline]
+    pub const fn new() -> MaybeVec<T> {
+        Self {
+            // use the system allocator instead of the global allocator
+            buf: RawVec::new_in(System),
+            len: 0,
+        }
+    }
+
+    /// Appends an element to the back of a collection.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the number of elements in the vector overflows a `usize`.
+    pub fn push(&mut self, value: T) {
+        // This will panic or abort if we would allocate > isize::MAX bytes
+        // or if the length increment would overflow for zero-sized types.
+        if self.len == self.buf.capacity() {
+            self.reserve(1);
+        }
+        unsafe {
+            let end = self.as_mut_ptr().add(self.len);
+            ::std::ptr::write(end, Some(value));
+            self.len += 1;
+        }
+    }
+
+    /// Reserves capacity for at least `additional` more elements to be inserted
+    /// in the given `MaybeVec<T>`. The collection may reserve more space to avoid
+    /// frequent reallocations. After calling `reserve`, capacity will be
+    /// greater than or equal to `self.len() + additional`. Does nothing if
+    /// capacity is already sufficient.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the new capacity overflows `usize`.
+    pub fn reserve(&mut self, additional: usize) {
+        self.buf.reserve(self.len, additional);
+    }
+
+    /// Removes and returns the element at position `index` within the vector if
+    /// it exists. Removal is O(1): the value at position `index` is simply
+    /// replaced with `None`. Since the vector does not shift all elements after
+    /// it to the left, removal does not decrease the length of the vector.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `index` is out of bounds.
+    pub fn remove(&mut self, index: usize) -> Option<T> {
+        assert!(index < self.len());
+        unsafe {
+            // infallible
+            let ret;
```

Comment:
I don't think the inner block does anything here -- the `unsafe` block itself defines a scope of equivalent length?

Reply:
This was inspired by `Vec`'s `remove` method which I chopped and changed in order to *not* shift. I'm guessing this is most likely a hangover from pre-NLL Rust. Simplified here: 3d7f208a62b8a86334878604d85d676bc4277712